### PR TITLE
fix(cursor): consolidate hook event dispatch

### LINKED ISSUE #2419: Cursor scaffold wires only sessionStart — port the profile-appropriate hook subset to the Cursor 3.x hook contract
## Summary

Follow-up to #2065 / #2066. That work gave Cursor an independent memory root (`ECC_AGENT_DATA_HOME`) and wired a single `sessionStart` hook (`cursor-session-env.js`). But the Cursor scaffold's `scaffolds/cursor/hooks.json` still registers **only that one hook**, while `scripts/hooks/` ships ~48 hook scripts that are wired for the Claude Code target. So an ECC-on-Cursor user today gets the data-root isolation but almost none of ECC's hook-driven behaviour — no continual-learning on stop, no post-edit format/typecheck, no quality gate, no commit-quality guard, no compaction hooks.

This proposes wiring the **generalizable subset** of ECC's existing hooks into the Cursor scaffold, adapted to Cursor's hook contract and gated by hook profile.

## Current state

`scaffolds/cursor/hooks.json`:

```json
{ "version": 1, "hooks": { "sessionStart": [ { "command": "node .cursor/scripts/hooks/cursor-session-env.js" } ] } }
```

## Cursor hook contract (3.x) — relevant facts

Cursor's agent hooks map closely onto ECC's Claude-side model, so most hooks port by event-name mapping + matcher translation. Proposed mapping:

| ECC hook(s) | Cursor event | Blocking |
|---|---|---|
| `post-edit-format` / `-typecheck` / `-console-warn` | `afterFileEdit` (matcher `Write`) | no |
| `pre-bash-commit-quality`, `block-no-verify`, `pre-bash-dev-server-block` | `beforeShellExecution` (regex matcher) | yes (permission) |
| `post-bash-*` | `afterShellExecution` | no |
| `quality-gate`, `stop-format-typecheck`, `evaluate-session` | `stop` | no |
| `pre-compact` / `suggest-compact` | `preCompact` | no |
| `session-end` / `-marker` | `sessionEnd` | no |

Four contract constraints the implementation must respect (all observed on current Cursor):

1. **One hook per event array.** Cursor currently executes only the first entry in an event's array. ECC already solves this on the Claude side with `*-dispatcher.js` fan-out scripts — the Cursor wiring should register a single dispatcher per event and fan out internally, not list N hooks.
2. **CLI vs editor surface.** The Cursor CLI does not emit every event the editor does; hooks should tolerate an event never firing on the CLI surface.
3. **`afterFileEdit` is not exhaustive** on batch/mid-session edits — post-edit hooks must be advisory, not relied on for correctness.
4. **Cloud-agent gaps.** Cloud agents do not receive `stop`, `sessionStart`/`sessionEnd`, or MCP events, and run only command-based hooks. Continual-learning-on-`stop` must degrade gracefully (no error) when the event never arrives.

`hooks.json` requires `"version": 1`.

## Proposed scope

- **Include** (generalizable, useful to any ECC-on-Cursor user): continual-learning (`stop`/`sessionEnd` → `evaluate-session`), post-edit format/typecheck, quality gate, commit-quality / `block-no-verify` shell guards, pre-compact.
- **Exclude** (environment- or operator-specific; keep out of the default scaffold): desktop notifications, tmux/dev-server automation, any vendor-specific security monitors, cost tracking.
- **Profile-gate** it: honour the existing `ECC_HOOK_PROFILE` (`minimal` → sessionStart + continual-learning only; `full` → the whole applicable subset). `minimal` should remain the safe default for Cursor.
- Prefer the dispatcher pattern so the "first-hook-only" limitation and future event additions are both handled in one place.

## Why this is worth doing upstream

The gap affects every ECC user who runs on Cursor, not any one setup — the hooks already exist; they're simply not wired for the Cursor target. Completing the parity that #2066 started makes the Cursor install a first-class citizen rather than a memory-root-only install.

## Offer

Happy to implement this if the direction is acceptable — including the Cursor-contract adapters, dispatcher wiring, and profile gating, tested against both the Cursor editor and CLI surfaces.

### PR DESCRIPTION
**Pull request by wellkilo**
## What Changed

- Consolidated every Cursor hook event behind a single event-level dispatcher, including the duplicate `beforeShellExecution` registrations.
- Added consistent master-switch, profile, and per-hook gating across Cursor-native hooks.
- Made the adapter resolve shared hook runtime files in both the repository and installed `.cursor/scripts` layouts.
- Mapped Cursor's official `file_path` and `conversation_id` fields into the shared hook contract, keeping concurrent edit accumulators session-scoped.
- Combined Cursor data-root setup with the shared SessionStart context hook, including safe malformed-input fallback.
- Made blocking pre-execution hooks fail closed on oversized UTF-8 input while respecting explicit hook disablement, and recognized Cursor's official `file_path` in sensitive-file guards.
- Added the batched format/typecheck hook to Cursor Stop with bounded time budgets while preserving lifecycle persistence first.
- Ensured `platform-configs` alone does not register hooks without their runtime; `hooks-runtime` now owns the native Cursor hook registry and dispatchers.
- Updated Cursor documentation and added contract/regression coverage for installed layouts, profiles, exit codes, malformed and oversized input, warnings, sessions, and Stop behavior.

## Why This Change

Cursor's registry still had two `beforeShellExecution` entries, incomplete profile controls, and no Stop-time format/typecheck batch. A single dispatcher per event provides deterministic ordering and remains compatible with Cursor versions that execute only the first matching entry.

Closes #2419

## Testing Done

- [x] Manual installation and reinstall testing completed
- [ ] Automated tests pass locally (`node tests/run-all.js`)
- [x] Edge cases considered and tested

Focused results:

- `node tests/hooks/cursor-hook-dispatchers.test.js` — 27 passed
- `node tests/hooks/cursor-block-no-verify.test.js` — 15 passed
- `node tests/hooks/stop-format-typecheck.test.js` — 25 passed
- `node tests/lib/install-targets.test.js` — 51 passed
- `node tests/lib/install-manifests.test.js` — 41 passed
- `node tests/lib/install-executor.test.js` — 16 passed
- `node tests/scripts/install-apply.test.js` — 40 passed
- Hook/install manifest/personal-path validators, focused ESLint, Markdown lint, and `git diff --check` passed

The full `npm test` suite was also attempted but is not clean in this sandbox (4,149 passed / 23 failed). Observed failures include permission errors writing outside the workspace and npm cache, the plan-canvas live-reload timeout, and existing validator fixture failures; the affected focused suites above pass.

## Type of Change

- [x] `fix:` Bug fix
- [ ] `feat:` New feature
- [ ] `refactor:` Code refactoring
- [x] `docs:` Documentation
- [x] `test:` Tests
- [ ] `chore:` Maintenance/tooling
- [ ] `ci:` CI/CD changes

## Security & Quality Checklist

- [x] No secrets or API keys committed (ghp_, sk-, AKIA, xoxb, xoxp patterns checked)
- [x] JSON files validate cleanly
- [x] Shell scripts pass shellcheck (not applicable; no shell scripts changed)
- [x] Pre-commit-equivalent focused checks pass locally
- [x] No sensitive data exposed in logs or output
- [x] Follows conventional commits format

## If you changed dependencies or `package.json` (`bin` / `files` / deps)

- [x] Not applicable; dependencies and package metadata were not changed.

## If you added a skill, command, agent, hook, or CLI tool

- [x] Not applicable; this consolidates and completes existing Cursor hook surfaces without adding a new catalog entry.

## Documentation

- [x] Updated relevant documentation
- [x] Added comments for complex logic
- [x] README updated

### CODE REVIEW

File: .cursor/hooks/adapter.js [hunk 1]
```diff
@@ -5,38 +5,109 @@
...
+      } else {
+        truncated = true;
+      }
+    });
+    process.stdin.on('end', () => {
+      if (truncated) {
+        process.stderr.write(
+          `[Cursor Hook] stdin exceeded ${MAX_STDIN} characters; suppressing truncated input\n`
+        );
+        resolve('');
```
[the remark this answers is not in the document]
Comment (wellkilo) [reply .cursor/hooks/adapter.js]: Fixed in 0718ec24. Input limits are now measured in UTF-8 bytes, and beforeShellExecution exits 2 on oversized payloads whenever either blocking shell guard is enabled. Explicit master/per-hook disablement remains honored. Regression coverage includes both ASCII and multibyte payloads.

File: .cursor/hooks/before-tab-file-read.js [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
-const { readStdin } = require('./adapter');
+const { readStdin, hookEnabled } = require('./adapter');
 readStdin().then(raw => {
+  if (!hookEnabled('pre:tab-read:sensitive-file-block', ['minimal', 'standard', 'strict'])) {
+    process.stdout.write(raw);
+    return;
+  }
   try {
     const input = JSON.parse(raw);
     const filePath = input.path || input.file || '';
```
[the remark this answers is not in the document]
Comment (wellkilo) [reply .cursor/hooks/before-tab-file-read.js]: Fixed in 0718ec24. Both sensitive-file hooks now prioritize Cursor's official file_path field, with a runtime regression test covering the warning and blocking paths.

File: .cursor/hooks/session-start.js [hunk 3]
```diff
@@ -1,10 +1,61 @@
...
+    process.stdout.write(JSON.stringify({}));
+    return;
+  }
+
+  const envResult = runExistingHook('cursor-session-env.js', raw, {
+    forwardStderr: true,
+  });
+  if (envResult && envResult.status === 0 && envResult.stdout) {
+    try {
+      outputs.push(normalizeSessionStartOutput(envResult.stdout));
```
[the remark this answers is not in the document]
Comment (wellkilo) [reply .cursor/hooks/session-start.js]: Addressed in 0718ec24. The SessionStart output accumulator now uses immutable array replacement while preserving output order.

File: .cursor/hooks/session-start.js [hunk 4]
```diff
@@ -1,10 +1,61 @@
...
+    } catch {
```
[the remark this answers is not in the document]
Comment (wellkilo) [reply .cursor/hooks/session-start.js:38]: Addressed in 0718ec24. Both normalization failure paths now emit fixed, non-sensitive stderr diagnostics while keeping SessionStart non-blocking.

File: .cursor/hooks/stop.js [hunk 5]
```diff
@@ -3,18 +3,36 @@ const { readStdin, runExistingHook, transformToClaude, hookEnabled } = require('
...
-    runExistingHook('evaluate-session.js', claudeInput);
+    runExistingHook('evaluate-session.js', claudeInput, sharedOptions);
   }
   if (hookEnabled('stop:cost-tracker', ['minimal', 'standard', 'strict'])) {
-    runExistingHook('cost-tracker.js', claudeInput);
+    runExistingHook('cost-tracker.js', claudeInput, sharedOptions);
+  }
+  if (hookEnabled('stop:format-typecheck', ['standard', 'strict'])) {
+    runExistingHook('stop-format-typecheck.js', claudeInput, {
+      timeout: 225000,
```
[the remark this answers is not in the document]
Comment (wellkilo) [reply .cursor/hooks/stop.js]: Addressed in 0718ec24. The format/typecheck budget and timeout headroom are shared constants, and the outer timeout is derived by createStopFormatTypecheckOptions. The test now verifies the observable helper contract.

File: scripts/hooks/stop-format-typecheck.js [hunk 6]
```diff
@@ -197,7 +207,8 @@ function main() {
   // Distribute the budget evenly across all batches so the cumulative total
   // stays within the Stop hook wall-clock limit even in large monorepos.
   const totalBatches = byProjectRoot.size + byTsConfigDir.size;
-  const perBatchMs = totalBatches > 0 ? Math.floor(TOTAL_BUDGET_MS / totalBatches) : 60_000;
+  const totalBudgetMs = getTotalBudgetMs();
+  const perBatchMs = totalBatches > 0 ? Math.floor(totalBudgetMs / totalBatches) : 60_000;
```
[the remark this answers is not in the document]
Comment (wellkilo) [reply scripts/hooks/stop-format-typecheck.js]: Fixed in 0718ec24. Per-batch timeouts are now clamped to at least 1 ms, with a regression test for totalBatches greater than totalBudgetMs.

File: tests/hooks/cursor-block-no-verify.test.js [hunk 7]
```diff
@@ -14,17 +14,22 @@ const assert = require('assert');
...
+  const result = spawnSync('node', [dispatcher], {
     input: rawInput,
     encoding: 'utf8',
-    env: { ...process.env, ECC_HOOK_PROFILE: 'standard', ...env },
+    env: {
+      ...process.env,
+      ECC_HOOK_PROFILE: 'standard',
+      ECC_DISABLED_HOOKS: 'pre:bash:dev-server-block',
+      ...env,
+    },
```
[the remark this answers is not in the document]
Comment (wellkilo) [reply tests/hooks/cursor-block-no-verify.test.js:33]: Fixed in 0718ec24. The block-no-verify test harness pins ECC_HOOKS_ENABLED=true so ambient environment state cannot disable its blocking assertions.

File: tests/hooks/cursor-hook-dispatchers.test.js [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+
+if (test('Stop reserves time for lifecycle hooks before the format batch', () => {
+  const source = fs.readFileSync(STOP, 'utf8');
+  assert.ok(
+    source.indexOf("runExistingHook('session-end.js'")
+      < source.indexOf("runExistingHook('stop-format-typecheck.js'"),
+    'Session persistence should run before the long format/typecheck batch'
+  );
+  assert.match(source, /timeout:\s*225000/);
+  assert.match(source, /ECC_STOP_FORMAT_TYPECHECK_BUDGET_MS:\s*'210000'/);
```
[the remark this answers is not in the document]
Comment (wellkilo) [reply tests/hooks/cursor-hook-dispatchers.test.js]: Addressed in 0718ec24. The Stop test now executes an installed fixture and verifies session-end runs before stop-format-typecheck, confirms the 210000 ms child budget, and checks the derived 225000 ms outer timeout through the helper contract.

File: .cursor/hooks/adapter.js [hunk 9]
```diff
@@ -5,38 +5,120 @@
...
+      }
+      resolve(options.includeMetadata === true
+        ? { raw: data, truncated: false }
+        : data);
+    });
+    process.stdin.on('error', error => {
+      process.stderr.write(`[Cursor Hook] stdin read failed: ${error.message}\n`);
+      resolve(options.includeMetadata === true
+        ? { raw: '', truncated: false }
+        : '');
```
[the remark this answers is not in the document]
Comment (wellkilo) [reply .cursor/hooks/adapter.js:52]: Fixed in 2de2a592. readStdin metadata now includes readError, and both blocking dispatchers treat readError or truncation as untrusted input and exit 2 while their guards are enabled. The regression injects a synthetic stream error and verifies the explicit marker; legacy callers still receive strings.

File: .cursor/hooks/before-tab-file-read.js [hunk 10]
```diff
@@ -1,13 +1,23 @@
...
+    return;
+  }
+  if (truncated) {
+    console.error('[Cursor Hook] stdin exceeded the safety limit; blocking beforeTabFileRead');
+    process.exit(2);
+  }
   try {
     const input = JSON.parse(raw);
-    const filePath = input.path || input.file || '';
+    const filePath = input.file_path || input.path || input.file || '';
```
[the remark this answers is not in the document]
Comment (wellkilo) [reply .cursor/hooks/before-tab-file-read.js]: Fixed in 2de2a592. Both sensitive-file hooks now select the first non-empty string among file_path, path, and file. Regression coverage verifies an object-valued file_path cannot shadow a valid .env fallback.
